### PR TITLE
source-microsoft-entra-id contribution from bishalbera

### LINKED ISSUE #46614: [connector-request] Microsoft Entra ID
This is a **suggestion** to create a connector for **Microsoft Entra ID** using the UI Connector Builder.
The website for the connector is https://www.microsoft.com/en-us/security/business/identity-access/microsoft-entra-id.

If you want to try, claim this issue and start working on it following the steps below.

### Steps:
1. Comment in the issue and wait to be assigned to start working on it.
2. Map the API endpoints (get approval to move to 3 step)
3. Describe steps to get credentials.
4. Create the connector using the UI Builder

> [!IMPORTANT]
> Airbyte doesn't have sandbox credentials for this connector. You must have access/credentials to the service provider to create the connector.

> [!CAUTION]
> This is a **suggestion** and there may be cases where creating the connector using the Builder won't be possible. To minimize the risk of investing a lot of time directly in creating the connector, we strongly recommend to follow steps above.

### PR DESCRIPTION
**Pull request by bishalbera**
## What

This PR adds a new connector for source Microsoft Entra Id (source-microsoft-entra-id).

Contributor-provided connector description:

The Microsoft Entra ID Connector for Airbyte allows seamless integration with Microsoft Entra ID, enabling secure and automated data synchronization of identity and access management information. With this connector, users can efficiently retrieve and manage user, group, and directory data to streamline identity workflows and ensure up-to-date access control within their applications.
## Details

<details><summary>Expand for list of spec properties and streams</summary>

### Configuration

| Input | Type | Description | Default Value |
|-------|------|-------------|---------------|
| `type` | `string` | Type.  |  |
| `client_id` | `string` | Client ID.  |  |
| `client_secret` | `string` | Client secret.  |  |
| `tenant_id` | `string` | Tenant Id.  |  |
| `application_id_uri` | `string` | Application Id URI.  |  |
| `user_id` | `string` | User Id.  |  |
| `ids` | `array` | Ids.  |  |
| `types` | `array` | Types.  |  |

### Streams
| Stream Name | Primary Key | Pagination | Supports Full Sync | Supports Incremental |
|-------------|-------------|------------|---------------------|----------------------|
| users | id | DefaultPaginator | ✅ |  ❌  |
| groups | id | DefaultPaginator | ✅ |  ❌  |
| applications | id | DefaultPaginator | ✅ |  ❌  |
| user_owned_deleted_items | id | DefaultPaginator | ✅ |  ❌  |
| directoryroles | id | No pagination | ✅ |  ❌  |
| auditlogs | id | DefaultPaginator | ✅ |  ❌  |
| directoryobjects | id | DefaultPaginator | ✅ |  ❌  |
| directoryroletemplates | id | No pagination | ✅ |  ❌  |
| directoryaudits | id | DefaultPaginator | ✅ |  ❌  |
| serviceprincipals | id | DefaultPaginator | ✅ |  ❌  |
| identityproviders |  | DefaultPaginator | ✅ |  ❌  |
| adminconsentrequestpolicy |  | DefaultPaginator | ✅ |  ❌  |

</details>

## Reviewer checklist
- [ ] Run `/format-fix` to resolve any formatting errors
- [ ] Help the contributor upload an icon that follows our icon design guidelines
- [ ] Click into the CI workflows that wait for a maintainer to run them, which should trigger CI runs

### CODE REVIEW

File: docs/integrations/sources/microsoft-entra-id.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+| `tenant_id` | `string` | Tenant Id.  |  |
+| `application_id_uri` | `string` | Application Id URI.  |  |
+| `user_id` | `string` | User Id.  |  |
+| `ids` | `array` | Ids.  |  |
+| `types` | `array` | Types.  |  |
+
+## Streams
+| Stream Name | Primary Key | Pagination | Supports Full Sync | Supports Incremental |
+|-------------|-------------|------------|---------------------|----------------------|
+| users | id | No pagination | ✅ |  ❌  |
```
Comment (marcosmarxm) [inline docs/integrations/sources/microsoft-entra-id.md:20]: Please check instructions how to implement pagination in Microsoft https://learn.microsoft.com/en-us/graph/best-practices-concept#pagination

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

File: docs/integrations/sources/microsoft-entra-id.md (hunk 1 above)
Comment (bishalbera) [reply docs/integrations/sources/microsoft-entra-id.md:20]: ok

Comment (bishalbera) [reply docs/integrations/sources/microsoft-entra-id.md]: @marcosmarxm added pagination . check plz .

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
The pagination is not correct. I tested using the `serviceprincipals` stream.
To simulate the pagination I injected the `$top` `5`
image


To fix the pagination please update the code to:
image

Cursor Value: `{{ response.get("@odata.nextLink") | regex_search('\$skiptoken=([^&]+)') }}`

Let me know when you update all streams!

**Comment by bishalbera** [comment]
> {{ response.get("@odata.nextLink") | regex_search('\$skiptoken=([^&]+)') }}

do i need to use the `top` parameter on every stream or was it only for your testing purpose. Because when I tried without specifying it its giving me error saying skiptoken is invalid
Screenshot 2024-10-18 at 11 11 48 AM

**Comment by bishalbera** [comment]
@marcosmarxm Update: I think, I've figured it out. It was giving error because when it couldn't find the value its taking the mentioned path as its value. So I've set the stopping condition to `{{ response.get("@odata.nextLink") is none }}`. Now the error is gone. I've pushing the changes. Please verify it. Thanks :)

**Comment by marcosmarxm** [comment]
Hello 👋 Due to the numerous contributions happening, I decided to create this [view](https://github.com/orgs/airbytehq/projects/104/views/3) to help you know when I plan to review your contribution. If your connector has some limited-time credential feel free to reach me out.

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**